samza-sql/src/main/java/org/apache/samza/sql/interfaces/SamzaRelConverter.java

```diff
@@ -0,0 +1,78 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.interfaces;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.schema.TableFactory;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+
+
+/**
+ * Samza SQL application uses {@link SamzaRelConverter} to convert the Samza messages to relational message before
+ * it can be processed by the calcite engine.
+ * The {@link SamzaRelConverter} is configurable at a system level, So it is possible to configure different
+ * {@link SamzaRelConverter} for different systems.
+ */
+public interface SamzaRelConverter {
+
+  /**
+   * Converts the schema to relational schema. It is possible that there can be multiple tables associated
+   * within a single schema.
+   * @param systemName the system name
+   * @param streamName Name of the stream
+   * @return Map of table name and it's associated relational schemas.
+   */
+  TableFactory<Table> getRelationalSchema(String systemName, String streamName);
+
+  /**
+   * Converts the object to relational message corresponding to the tableName with relational schema.
+   * @param key Key corresponding to the message.
+   * @param message message object that needs to be converted.
+   * @param dataType relational schema associated to the table.
+   * @return Relational message extracted from the object.
+   */
+  SamzaSqlRelMessage convertToRelMessage(Object key, Object message, RelDataType dataType);
```

Comment:
Can this be typed? we have a KV<K, V> type to represent the key, value pair. Can we use it here? 

Another question is that whether we can infer the RelDataType if the KV is typed.